<a href="https://colab.research.google.com/github/bkloner/QTM150-Final/blob/master/Copy_of_HW1AnswersQTM350.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Question 1

##Member 1: **Jon Marcus**


My name is Jon Marcus. I am a junior at Emory studying QSS + Informatics. I'm from Fair Lawn, NJ, about 20 minutes from NYC. I am a huge sports fan, and it was the statistical side of baseball that got me interested in studying data science. I am excited to learn more about this class so I can see what being a data scientist truly means. [Here](https://www.linkedin.com/in/jon-marcus-1b8197180/) is my linkedin profile. 

In [ ]:
from IPython.display import display, Javascript
from google.colab.output import eval_js
from base64 import b64decode

def take_photo(filename='photo.jpg', quality=0.8):
  js = Javascript('''
    async function takePhoto(quality) {
      const div = document.createElement('div');
      const capture = document.createElement('button');
      capture.textContent = 'Capture';
      div.appendChild(capture);

      const video = document.createElement('video');
      video.style.display = 'block';
      const stream = await navigator.mediaDevices.getUserMedia({video: true});

      document.body.appendChild(div);
      div.appendChild(video);
      video.srcObject = stream;
      await video.play();

      // Resize the output to fit the video element.
      google.colab.output.setIframeHeight(document.documentElement.scrollHeight, true);

      // Wait for Capture to be clicked.
      await new Promise((resolve) => capture.onclick = resolve);

      const canvas = document.createElement('canvas');
      canvas.width = video.videoWidth;
      canvas.height = video.videoHeight;
      canvas.getContext('2d').drawImage(video, 0, 0);
      stream.getVideoTracks()[0].stop();
      div.remove();
      return canvas.toDataURL('image/jpeg', quality);
    }
    ''')
  display(js)
  data = eval_js('takePhoto({})'.format(quality))
  binary = b64decode(data.split(',')[1])
  with open(filename, 'wb') as f:
    f.write(binary)
  return filename

In [ ]:
from IPython.display import Image
try:
  filename = take_photo()
  print('Saved to {}'.format(filename))
  
  # Show the image which was just taken.
  display(Image(filename))
except Exception as err:
  # Errors will be thrown if the user does not have a webcam or if they do not
  # grant the page permission to access it.
  print(str(err))

<IPython.core.display.Javascript object>

NotAllowedError: Permission denied


##Member 2: **Vicky Liu**

My name is Vicky Liu. I am a junior at Emory University, double majoring in Physics and QSS on a Biology track. I was born in Brooklyn, New York, but grew up in Potomac, Maryland. 

I have linked my Linkedin profile [here](https://www.linkedin.com/in/vickyliu736/). 

##Member 3: **Jonathan Savdie**


Hi! My name is Jonathan Savdie, I am a junior majoring in QSS and my track is Economics. I was originally born in Colombia but I live in Costa Rica. I aspire to learn more about cloud computing and how it is being used in the ever-evolving world of data science. In my free time I like watching sports; especially soccer, football and mixed martial arts.

You can find my LinkedIn profile [here](https://www.linkedin.com/in/jonathan-s-33958911a)

## Member 4: **Brett Kloner**

My name is Brett Kloner, and I am a BBA + QSS major, concentrating in Finance. Ever since I took AP Stats in High School, I knew that I wanted to study some form of statistics and data science along with anything else I decided to do. I am really excited to learn Cloud Computing at the college level, so that I can take it with me into the workforce. In my free time, I like to play and watch baseball.
You can find my LinkedIn profile [here](https://www.linkedin.com/in/brett-kloner-b19645186/)

##Member 5: **Bryant Wise**

> My name is Bryant Wise. I am a junior on the baseball team here at Emory and I am studying QSS with a focus in Economics. Throughout my time at Emory, I have noticed how prevelant coding is in today's world, and this is why I am very excited to learn about Cloud Computing. You can find my LinkedIn profile [here](https://www.linkedin.com/in/bryant-wise-9698b2196/)


#Question 2

$$ \log(e^x)= x$$
Using the above equation, what is the smallest integer that fails?

In [ ]:
import numpy as np
np.log((np.e)**(-746))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: divide by zero encountered in log
  


-inf

Numpy fails at x=-746. This is because when you raise something to a negative power, it puts it in the denominator of a fraction. When the power is large enough, numpy essentially interprets the number as 0. When numpy tries to calculate the log of 0, it gets an error, since you cannot compute the log of 0. 

#Question 3

$$r = \frac{\log(V/P)}{t}$$ 

Using the formula above and the given values, we are able to calculate the rate that tuition increases per year, if it were compounded continuously. We then multiply the answer we get by 100 to get it into the form of a percentage. Below are the calcuations for six of the Ivy League schools, which we decided on because as Ivy League schools, we expect them to attract the same type of student, and therefore should see a similar rate of tuition increase.

##Harvard
(data was given)

In [ ]:
import numpy as np
Pnew=4070
Vnew=46340
tnew=2019-1970
rnew = np.log(Vnew/Pnew)/tnew
percent = rnew*100
f"So the annual rate, compounded continuously, is {round(percent, 2)}% for Harvard between 1970 and 2019."

'So the annual rate, compounded continuously, is 4.96% for Harvard between 1970 and 2019.'

##Cornell
(data found [here](https://www.nytimes.com/1970/01/25/archives/cornell-tuition-and-fees-to-increase-275-a-year.html) and [here](https://www.collegetuitioncompare.com/edu/190415/cornell-university/tuition/))

In [ ]:
Pnew=2600
Vnew=57222
tnew=2019-1970
rnew = np.log(Vnew/Pnew)/tnew
percent = rnew*100
f"So the annual rate, compounded continuously, is {round(percent, 2)}% for Cornell between 1970 and 2019."

'So the annual rate, compounded continuously, is 6.31% for Cornell between 1970 and 2019.'

##Yale
(data found [here](https://yaledailynews.com/blog/2005/04/13/ballooning-tuition-presents-challenges/#:~:text=Yale%20tuition%20had%20grown%20to,1980%2C%20according%20to%20Wolfe's%20figures.) and [here](https://www.cnbc.com/2019/04/18/it-costs-75925-to-go-to-yaleheres-how-much-students-actually-pay.html#:~:text=Yale%2C%20a%20prestigious%20member%20of,a%20sum%20total%20of%20%2472%2C100.))

In [ ]:
Pnew=2550
Vnew=55500
tnew=2019-1970
rnew = np.log(Vnew/Pnew)/tnew
percent = rnew*100
f"So the annual rate, compounded continuously, is {round(percent, 2)}% for Yale between 1970 and 2019."

'So the annual rate, compounded continuously, is 6.29% for Yale between 1970 and 2019.'

##University of Pennsylvania
(data found [here](https://archives.upenn.edu/exhibits/penn-history/tuition/tuition-1970-1979) and [here](https://penntoday.upenn.edu/news/penn-trustees-approve-2019-20-financial-aid-budget-and-tuition#:~:text=To%20maintain%20Penn's%20world%2Dclass,housing%2C%20and%20%245%2C590%20for%20dining.))

In [ ]:
Pnew=2350
Vnew=51156
tnew=2019-1970
rnew = np.log(Vnew/Pnew)/tnew
percent = rnew*100
f"So the annual rate, compounded continuously, is {round(percent, 2)}% for the University of Pennsylvania between 1970 and 2019."

'So the annual rate, compounded continuously, is 6.29% for the University of Pennsylvania between 1970 and 2019.'

##Princeton
(data from [here](https://www.nytimes.com/1973/02/19/archives/princeton-raising-tuition-again-and-it-sees-no-end-to-increases.html) and [here](https://www.cnbc.com/2019/05/03/it-costs-73450-to-go-to-princetonheres-how-much-students-pay.html#:~:text=According%20to%20the%20Princeton%20undergraduate,a%20%24930%20residential%20college%20fee.))

In [ ]:
Pnew=3300
Vnew=51870
tnew=2019-1973
rnew = np.log(Vnew/Pnew)/tnew
percent = rnew*100
f"So the annual rate, compounded continuously, is {round(percent, 2)}% for Princeton between 1973 and 2019."

'So the annual rate, compounded continuously, is 5.99% for Princeton between 1973 and 2019.'

##Columbia
(data from [here](https://www.nytimes.com/1985/06/01/us/college-cost-has-matched-inflation-since-74.html) and [here](https://www.univstats.com/colleges/columbia-university-in-the-city-of-new-york/cost-of-attendance/#:~:text=For%20the%20academic%20year%202019,a%20student%20lives%20off%2Dcampus.))

In [ ]:
Pnew=3370
Vnew=60578
tnew=2019-1974
rnew = np.log(Vnew/Pnew)/tnew
percent = rnew*100
f"So the annual rate, compounded continuously, is {round(percent, 2)}% for Columbia between 1974 and 2019."

'So the annual rate, compounded continuously, is 6.42% for Columbia between 1974 and 2019.'

In conclusion, Harvard has had a smaller annual tuition increase on average compared to the other Ivy League schools. Columbia, Princeton, University of Pennsylvania, Yale, and Cornell all had higher annual increases, but it is important to note that in the 1970s, Harvard's tuition was slightly higher than their ivy counterparts, so their tuition increased at a slower rate to keep the tuition prices of the Ivy League schools relatively similar. 

#Question 4

To get the day of the week we were born, we first save our date of birth as a date, using pandas. Then we can simply use the function that gives the day of the week from that respective date.

###Jon Marcus

In [ ]:
import pandas as pd
jondob = pd.Timestamp('May 11, 2000')
jondob.day_name()

'Thursday'

###Vicky Liu

In [ ]:
vickydob = pd.Timestamp('September 6, 2000')
vickydob.day_name()

'Wednesday'

###Jonathan Savdie

In [ ]:
jonathandob = pd.Timestamp('February 7, 2000')
jonathandob.day_name()

'Monday'

###Brett Kloner

In [ ]:
brettdob = pd.Timestamp('March 13, 2000')
brettdob.day_name()

'Monday'

###Bryant Wise

In [ ]:
bryantdob = pd.Timestamp('August 9, 2000')
bryantdob.day_name()

'Wednesday'

###Days since our birthdays

To get the days since our birthdays, we created a timestamp for "now", and subtracted our birthdays from it, to get the amount of days since we were born. We then combined each of the answers we got into one variable, $totdays$. 

In [ ]:
now = pd.Timestamp.now()
daysjon = now-jondob
daysvicky = now-vickydob
daysjonathan = now-jonathandob
daysbrett = now-brettdob
daysbryant = now-bryantdob
totdays = daysjon, daysvicky, daysjonathan, daysbrett, daysbryant
totdays

(Timedelta('7582 days 03:07:08.424928'),
 Timedelta('7464 days 03:07:08.424928'),
 Timedelta('7676 days 03:07:08.424928'),
 Timedelta('7641 days 03:07:08.424928'),
 Timedelta('7492 days 03:07:08.424928'))

Bryant has had the fewest days since his birthday, while Jonathan has had the most.

##Question 5

In [ ]:
#@title Interest rate on 30 year mortgage (e.g. if 3%, enter 3)
percentage =  2#@param {type:"number"}
rate = percentage/100
rate

0.02

In [ ]:
house_spend=63179/4
affordable_monthly_payment = house_spend/12
affordable_monthly_payment
c=rate/12 # monthly interest rate
n = 12*30 # number of months in 30 years

In [ ]:
import numpy as np

$$ saver house price = \frac{\ house spend ((1 + rate)^{time-1})}{rate}$$ 

How many years will it take for the saver profile buyer to save enough to purchase the $312,196 home that the renter profile can afford?

Task 1:
First off, we have to change the value of the saver house price to 312,196 USD as it says in the givens. To avoid confusion, we just created a new variable called "new_saver_house_price". Next, we rearranged the saver_house_price formula above in order to set the variables in terms of Timw. Since doing so reqiured logs, we figured it would be easier to make the numerator and denominator their own variables.

In [ ]:
#task1
new_saver_house_price = 312196    # price of home
numerator = np.log(1+(new_saver_house_price*rate)/house_spend)  # equation from above
denominator = np.log(1+rate)
T = (numerator/denominator)   # to find T: time in years
T

16.822062361663896

From the code above, we see that if the someone wants to save and be able to pay everything uprfront for a 312,196 USD house, they will have to put aside the required 1/4 of their income (assuming it is 63,179 USD) for 16.82 years before they can do so.

Calculate what how many years the saver profile needs to save to be able to buy a house which is priced at the median price for homes in Atlanta.

Task 2:
This task requires the same work as the task above, just with a slightly different house price of 308,663 USD. Therefore, we set it up the same way, just creating and using different variables such as median_saver_house_price, numerator2, denominator2, and T2. We keep rate and house_spend the same.

In [ ]:
#task2
median_saver_house_price = 308663
numerator2 = np.log(1+(median_saver_house_price*rate)/house_spend)
denominator2 = np.log(1+rate)
T2 = (numerator2/denominator2)
T2

16.65989532416991

From the code above, we see that if the someone wants to save and be able to pay everything uprfront for a 308,663 USD house, which happens to be the median price of a house in Atlanta, they will have to put aside the required 1/4 of their income (assuming it is 63,179 USD) for 16.66 years before they can do so.

> What if the interest rate is 3% instead of 2%?





Task 3:
In the next two coding cells, we are doing the same work as in task 2, except that we are changing the interest rates to see how it affects how long a person making the median income will have to save before they can pay for a median priced house in Atlanta. As a result, median_saver_house_price stays the same, while we change rate, and therefore use the new rate variables (rate3 and rate3b) in the numerator and denominator formulas, which then caused us to make new numerator, denominator, and T variables (numerator 3 and 3b, denominator 3 and 3b, T3 and 3b).

In [ ]:
#task3
median_saver_house_price = 308663
rate3 = 0.03
numerator3 = np.log(1+(median_saver_house_price*rate3)/house_spend)
denominator3 = np.log(1+rate3)
T3 = (numerator3/denominator3)
T3

15.608935362586246

Increasing the interest rate from 2 percent to 3 percent makes it so that a person making the median income and putting 1/4 of it aside per year would have to wait 15.61 years to buy the median price of a house in Atlanta, which is one year sooner than with a 2 percent interest rate.

> What if it is 10% as it was in the 1980's?

In [ ]:
#task3
median_saver_house_price = 308663
rate3b = 0.1
numerator3b = np.log(1+(median_saver_house_price*rate3b)/house_spend)
denominator3b = np.log(1+rate3b)
T3b = (numerator3b/denominator3b)
T3b

11.365335428425023

Increasing the interest rate from 2 percent to 10 percent makes it so that a person making the median income and putting 1/4 of it aside per year would have to wait 11.37 years to buy the median price of a house in Atlanta, which is five years sooner than with a 2 percent interest rate.

Suppose that a house is currently on the market with an asking price of $725000 when the interest rate is 3%. Calculate how large house_spend would need to be for the renter profile buyer to be able to purchase this house.

Task 4:
First, we create the variable rate4 to represent the 3 percent interest rate, and as a result create a new variable c4 to include rate4. Moreover, we create new_renter_house_price to represent the 725,000 USD home. Next, we rearrange the renter_house_price formula above so that Affordable Monthly Payment is isolated, and represented as new_amp with c4 and new_renter_house_price now included in the formula. Lastly, the new house spend is equal to 12 times new_amp since house spend is yearly, and affordable monthly payment is monthly.

In [ ]:
#task4
rate4 = 0.03
c4=rate4/12 
n = 12*30
new_renter_house_price = 725000
new_amp= new_renter_house_price*((c4*(1 + c4)**n)/((1 + c4)**n - 1))
new_house_spend = new_amp*12
new_house_spend

36679.55093446267

In order for the renter to afford $725,000 on loans, he will have to put 36,670.55 USD towards it each year for 30 years.

> Suppose interest rates suddenly climb from 3 up to 4%. Calculate how large house_spend would need to be for the renter profile buyer to be able to purchase this house.

Task 5: 
This task requires the same work as Task 5, the only thing we are changing is in interest rate to 4 percent, and creating the variable rate5 to represent that. Additionally, we create new variables c5, new5_renter_house_price, new5_amp, and new5_house_spend to take into account the new interest rate.

In [ ]:
#task5
rate5 = 0.04
c5=rate5/12 
n = 12*30
new5_renter_house_price = 725000
new5_amp= new5_renter_house_price*((c5*(1 + c5)**n)/((1 + c5)**n - 1))
new5_house_spend = new5_amp*12
new5_house_spend

41535.13070549448

With interest rates now at 4 percent, the renter will now have to contribute 41535.13 USD per year for 30 years to pay off the 725,000 USD house on loans.

Suppose that suddenly interest rates drop to 2% from 4%. What can the renter afford to pay now? Should the home sellers raise the asking price from $725000 to a higher number?

Task 6:
We are changing the interest rate to 2 percent, and creating the variable rate6, and subsequently c6 and new6_renter_house_price to represent that. However, we are assuming the same monthly payment as task 5, and therefore keep the variable new5_amp. We now want to go back to isolating renter house price, like in the renter house price in the question givens/preface, so that we can see what the renter can now afford.

In [ ]:
#task6
rate6 = 0.02
c6=rate6/12
n = 12*30
new6_renter_house_price = new5_amp/((c6*(1 + c6)**n)/((1 + c6)**n - 1))
new6_renter_house_price

936438.9995325283

The renter can now afford a house of 936,439.00 USD, and therefore the seller should increase their asking price.

Bonus:
As shown in tasks 2 and 3, the least years required before a saver could buy a median priced home decrease as the interest rate decreases. Therefore, the late 1970s, due to incredibly high interest rates, would have seen the least years required for the saver to wait to buy a median priced home. To the contrary, as shown in task 6, the renter is able to purchase a more expensive home when in interest rates are lower. Thus, the renter would be able to afford the most expsenvie home possible, given a certain house spend and income, now, or in the lat 2000s and early 2010s.